rotate_test.go

```diff
@@ -322,6 +343,394 @@ func TestStepSetSecret(t *testing.T) {
+			case "1":
+				return expectedUser, "p1"
+			case "2":
+				return expectedUser, expectedSecret
+			case "0":
+				return expectedUser, previousSecret
+			default:
+				return "credUser", "credPass"
+			}
+		},
+	}
+	verifiedNewPassword := []string{}
+	verifiedOldPassword := []string{}
+	expectedNewPassword := []string{"p2", "p1", "p0"}
+	expectedOldPassword := []string{"p1", "p1", "p0"}
+
+	var gotUsername, gotPassword string
+	ps := test.MockPasswordSetter{
+		SetPasswordFunc: func(ctx context.Context, creds db.NewPassword) error {
+			gotUsername = creds.New.Username
+			gotPassword = creds.New.Password
+			return nil
+		},
+		VerifyPasswordFunc: func(ctx context.Context, creds db.NewPassword) error {
+			verifiedNewPassword = append(verifiedNewPassword, creds.New.Password)
+			verifiedOldPassword = append(verifiedOldPassword, creds.Current.Password)
+			if creds.Current.Password != previousSecret {
+				return fmt.Errorf("expected to error")
+			}
+			return nil
+		},
+	}
+
+	// Create a new Rotator to test
+	r := rotate.NewRotator(rotate.Config{
+		SecretsManager: sm,
+		SecretSetter:   ss,
+		PasswordSetter: ps,
+	})
+
+	// Simulate setSecret event from Secrets Manager
+	event := map[string]string{
+		"ClientRequestToken": "abc",
+		"SecretId":           "def",
+		"Step":               "setSecret",
+	}
+	_, err := r.Handler(context.TODO(), event)
+	if err != nil {
+		t.Error(err)
+	}
+
+	if gotUsername != expectedUser {
+		t.Errorf("got username %s, expected \"%v\"", gotUsername, expectedUser)
+	}
+	if gotPassword != expectedSecret {
+		t.Errorf("got password %s, expected \"%v\"", gotPassword, expectedSecret)
+	}
+	if diff := deep.Equal(expectedNewPassword, verifiedNewPassword); diff != nil {
+		t.Errorf("unexpected diff in new password verification: %v", diff)
+	}
+
+	if diff := deep.Equal(expectedOldPassword, verifiedOldPassword); diff != nil {
+		t.Errorf("unexpected diff in current password verification: %v", diff)
+	}
+
+	// The code should not call UpdateSecretVersionStage. That doesn't happen
+	// until the last step.
+	if updateSecretVersionCalled {
+		t.Errorf("UpdateSecretVersionStage called, expected no call by CreateSecret")
+	}
+
+	// ----------------------------------------------------------------------
+
+	// The code caches secrets by stage to save money on AWS API calls.
+	// We should have the first 2 calls from above:
+	if nCallsToGetSecretValue != 3 {
+		t.Errorf("GetSecretValue called %d times, expected 2", nCallsToGetSecretValue)
+	}
+
+	// Then if run the step again, the number of calls to GetSecretValue
+	// should not increase because the code uses cached values:
+	_, err = r.Handler(context.TODO(), event)
+	if err != nil {
+		t.Error(err)
+	}
+	if nCallsToGetSecretValue != 6 {
+		t.Errorf("GetSecretValue called %d times, expected 4", nCallsToGetSecretValue)
+	}
+}
+
+func TestStepSetSecret_DBMismatch(t *testing.T) {
+	// Test that all versions of secrets are mismatched and db is out of sync with secret manager.
```

Comment:
What's the expected result in this case? If they're all mismatched... then we rollback to PREVIOUS?

Reply:
Updated comments.  expect SetSecret to return error before starting SetPassword